Error with `is` assertion when using `exactOptionalPropertyTypes` and indexed fields from another interface

# Bug Report

### 🔎 Search Terms

exactOptionalPropertyTypes
is
assertion

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/JYOwLgpgTgZghgYwgAgCoQM5mQbwFDLKRYCeA-AFzJZSgDmA3HgL556iSyIoDyADmGAB7EBlwEimMFXRYA2gHJiYEgoC6TQgHotyEEOTQoQqMgDuACwghkAEyH0iV5DCEAbN0LP0JOyVioaek1kPxNfXWVKajBaEEYWNlsIBDc4KBQYAFcQBEERZFBUrOSMAEEPACUIAEcs4AzbfnzRAAohAWFRKgAFdME4NwAeZq6MAD4ASioOlrFgMVGRDCYgA)

### 💻 Code

```ts
interface Test {
  testy?: string;
}

interface Options {
  test: Test['testy'];
  // no error when doing the following
  // test: string;
  // or
  // test?: string;
}

declare function includesAllRequiredOptions(options: Partial<Options>): options is Options;
```

with tsconfig option `"exactOptionalPropertyTypes": true`

### 🙁 Actual behavior

Error:
```txt
src/test.ts:15:84 - error TS2677: A type predicate's type must be assignable to its parameter's type.
  Type 'Options' is not assignable to type 'Partial<Options>' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
    Types of property 'test' are incompatible.
      Type 'string | undefined' is not assignable to type 'string'.
        Type 'undefined' is not assignable to type 'string'.

15 declare function includesAllRequiredOptions(options: Partial<Options>): options is Options;
```

### 🙂 Expected behavior

No error

---

i think this *may* be related to https://github.com/microsoft/TypeScript/issues/51638, though in this case it also happens on typescript 4.8 and nightly

Reproduction Repository & Branch: https://github.com/typegoose/typegoose-testing/tree/typescript49ErrorAssertion

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -16855,7 +16855,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
                 const propType = getTypeOfSymbol(prop);
                 return accessExpression && getAssignmentTargetKind(accessExpression) !== AssignmentKind.Definite ? getFlowTypeOfReference(accessExpression, propType) :
-                    accessNode && isIndexedAccessTypeNode(accessNode) ? mapType(propType, type => type === missingType ? undefinedType : type) :
+                    accessNode && isIndexedAccessTypeNode(accessNode) && containsMissingType(propType) ? getUnionType([propType, undefinedType]) :
                     propType;
             }
             if (everyType(objectType, isTupleType) && isNumericLiteralName(propName)) {

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -16856,6 +16856,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 const propType = getTypeOfSymbol(prop);
                 return accessExpression && getAssignmentTargetKind(accessExpression) !== AssignmentKind.Definite ?
                     getFlowTypeOfReference(accessExpression, propType) :
+                    accessNode && isIndexedAccessTypeNode(accessNode) ?
+                    mapType(propType, type => type === missingType ? undefinedType : type) :
                     propType;
             }
             if (everyType(objectType, isTupleType) && isNumericLiteralName(propName)) {

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -16854,10 +16854,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     }
                 }
                 const propType = getTypeOfSymbol(prop);
-                return accessExpression && getAssignmentTargetKind(accessExpression) !== AssignmentKind.Definite ?
-                    getFlowTypeOfReference(accessExpression, propType) :
-                    accessNode && isIndexedAccessTypeNode(accessNode) ?
-                    mapType(propType, type => type === missingType ? undefinedType : type) :
+                return accessExpression && getAssignmentTargetKind(accessExpression) !== AssignmentKind.Definite ? getFlowTypeOfReference(accessExpression, propType) :
+                    accessNode && isIndexedAccessTypeNode(accessNode) ? mapType(propType, type => type === missingType ? undefinedType : type) :
                     propType;
             }
             if (everyType(objectType, isTupleType) && isNumericLiteralName(propName)) {

--- a/tests/baselines/reference/typePredicateAcceptingPartialOfRefinedType.symbols
+++ b/tests/baselines/reference/typePredicateAcceptingPartialOfRefinedType.symbols
@@ -1,0 +1,26 @@
+=== tests/cases/compiler/typePredicateAcceptingPartialOfRefinedType.ts ===
+// repro #51953
+
+interface Test {
+>Test : Symbol(Test, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 0, 0))
+
+  testy?: string;
+>testy : Symbol(Test.testy, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 2, 16))
+}
+
+interface Options {
+>Options : Symbol(Options, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 4, 1))
+
+  test: Test['testy'];
+>test : Symbol(Options.test, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 6, 19))
+>Test : Symbol(Test, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 0, 0))
+}
+
+declare function includesAllRequiredOptions(options: Partial<Options>): options is Options;
+>includesAllRequiredOptions : Symbol(includesAllRequiredOptions, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 8, 1))
+>options : Symbol(options, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 10, 44))
+>Partial : Symbol(Partial, Decl(lib.es5.d.ts, --, --))
+>Options : Symbol(Options, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 4, 1))
+>options : Symbol(options, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 10, 44))
+>Options : Symbol(Options, Decl(typePredicateAcceptingPartialOfRefinedType.ts, 4, 1))
+

--- a/tests/baselines/reference/typePredicateAcceptingPartialOfRefinedType.types
+++ b/tests/baselines/reference/typePredicateAcceptingPartialOfRefinedType.types
@@ -1,0 +1,17 @@
+=== tests/cases/compiler/typePredicateAcceptingPartialOfRefinedType.ts ===
+// repro #51953
+
+interface Test {
+  testy?: string;
+>testy : string | undefined
+}
+
+interface Options {
+  test: Test['testy'];
+>test : string | undefined
+}
+
+declare function includesAllRequiredOptions(options: Partial<Options>): options is Options;
+>includesAllRequiredOptions : (options: Partial<Options>) => options is Options
+>options : Partial<Options>
+

--- a/tests/cases/compiler/typePredicateAcceptingPartialOfRefinedType.ts
+++ b/tests/cases/compiler/typePredicateAcceptingPartialOfRefinedType.ts
@@ -1,0 +1,15 @@
+// @strict: true
+// @exactOptionalPropertyTypes: true
+// @noEmit: true
+
+// repro #51953
+
+interface Test {
+  testy?: string;
+}
+
+interface Options {
+  test: Test['testy'];
+}
+
+declare function includesAllRequiredOptions(options: Partial<Options>): options is Options;

